### PR TITLE
feat: Stake manager for 2.0 staking

### PR DESCRIPTION
**Pull request by fb-alexcq**
- Support woo staking 
- Support MP and staking 
- Compound
- Support multiple rewards

### CODE REVIEW

**Review by merlincdj — APPROVED** [review]

File: contracts/rewarders/MpRewarder.sol [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import "./BaseRewarder.sol";
+
+import "../interfaces/IRewardBooster.sol";
+
+contract MpRewarder is BaseRewarder {
```
[the remark this answers is not in the document]
Comment (fb-alexcq) [reply contracts/rewarders/MpRewarder.sol:8]: 50%, mp reward (base)

staked woo = 100,  APR 50% ,  一年 50 mp -->. rewardPerSec , 

lastRewardTs

File: contracts/rewarders/BaseRewarder.sol [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    mapping(address => uint256) public rewardDebt; // reward debt
+    mapping(address => uint256) public rewardClaimable; // shadow harvested reward
+
+    uint256 public accTokenPerShare;
+
+    IWooStakingManager public stakingManager;
+
+    address public immutable rewardToken; // reward token
+    uint256 public rewardPerBlock; // emission rate of reward
```
Comment (fb-alexcq) [inline contracts/rewarders/BaseRewarder.sol:28]: rewardPerSec
lastRewardTs

File: contracts/WooStakingManager.sol [hunk 3]
```diff
@@ -181,11 +180,13 @@ contract WooStakingManager is IWooStakingManager, BaseAdminOperation {
...
         // claim auto updates the reward for the user
-        uint256 amount = mpRewarder.claim(_user, address(this));
+        mpRewarder.claim(_user, address(this));
+    }
 
-        // NO need to transfer MP token to self again
-        mpBalance[_user] += amount;
-        mpTotalBalance += amount;
+    function addMP(address _user, uint256 _amount) public {
+        require(address(mpRewarder) == msg.sender, "RESTRICTED_TO_MP_REWARDER");
```
Comment (fb-alexcq) [inline contracts/WooStakingManager.sol:187]: Sorry to re-iterate. Better to push this part in a modifier:  "onlyMpRewarder"

File: test/typescript/MpRewarder.test.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    let mpToken: Contract;
+
+    beforeEach(async () => {
+        const signers = await ethers.getSigners();
+        owner = signers[0];
+        baseToken = signers[1];
+        user = signers[2];
+        user1 = signers[3];
+        user2 = signers[4];
```
Comment (fb-alexcq) [inline test/typescript/MpRewarder.test.ts:67]: Better to follow the test code of StakingManager

Comment (Cphilo) [reply test/typescript/MpRewarder.test.ts]: resolved

File: contracts/WooStakingManager.sol (hunk 3 above)
Comment (Cphilo) [reply contracts/WooStakingManager.sol]: resolved

**Review by fb-alexcq — COMMENTED** [review]
Ok, looks good.

**Pull request merged**